ambari-server/src/main/java/org/apache/ambari/server/checks/MissingOsInRepoVersionCheck.java

```diff
@@ -0,0 +1,92 @@
+ *  * to you under the Apache License, Version 2.0 (the
+ *  * "License"); you may not use this file except in compliance
+ *  * with the License.  You may obtain a copy of the License at
+ *  *
+ *  *     http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing, software
+ *  * distributed under the License is distributed on an "AS IS" BASIS,
+ *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  * See the License for the specific language governing permissions and
+ *  * limitations under the License.
+ *
+ */
+
+package org.apache.ambari.server.checks;
+
+import static java.util.stream.Collectors.toSet;
+import static org.apache.ambari.server.state.MaintenanceState.OFF;
+
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+import org.apache.ambari.server.AmbariException;
+import org.apache.ambari.server.controller.PrereqCheckRequest;
+import org.apache.ambari.server.orm.entities.RepoOsEntity;
+import org.apache.ambari.server.state.Cluster;
+import org.apache.ambari.server.state.Host;
+import org.apache.ambari.server.state.stack.PrereqCheckStatus;
+import org.apache.ambari.server.state.stack.PrerequisiteCheck;
+import org.apache.ambari.server.state.stack.upgrade.UpgradeType;
+
+import com.google.inject.Singleton;
+
+/**
+ * This checks if the source and target version has an entry for each OS type in the cluster.
+ */
+@Singleton
+@UpgradeCheck(
+  group = UpgradeCheckGroup.REPOSITORY_VERSION,
+  required = { UpgradeType.NON_ROLLING })
+public class MissingOsInRepoVersionCheck extends AbstractCheckDescriptor {
+  public static final String SOURCE_OS = "source_os";
+  public static final String TARGET_OS = "target_os";
+
+  public MissingOsInRepoVersionCheck() {
+    super(CheckDescription.MISSING_OS_IN_REPO_VERSION);
+  }
+
+  @Override
+  public void perform(PrerequisiteCheck prerequisiteCheck, PrereqCheckRequest request) throws AmbariException {
+    Set<String> osFamiliesInCluster = osFamiliesInCluster(cluster(prerequisiteCheck));
+    if (!targetOsFamilies(request).containsAll(osFamiliesInCluster)) {
+      prerequisiteCheck.setFailReason(getFailReason(TARGET_OS, prerequisiteCheck, request));
+      prerequisiteCheck.setStatus(PrereqCheckStatus.FAIL);
+      prerequisiteCheck.setFailedOn(new LinkedHashSet<>(osFamiliesInCluster));
+    } else if (!sourceOsFamilies(request).containsAll(osFamiliesInCluster)) {
+      prerequisiteCheck.setFailReason(getFailReason(SOURCE_OS, prerequisiteCheck, request));
+      prerequisiteCheck.setStatus(PrereqCheckStatus.FAIL);
+      prerequisiteCheck.setFailedOn(new LinkedHashSet<>(osFamiliesInCluster));
+    }
+  }
+
+  private Cluster cluster(PrerequisiteCheck prerequisiteCheck) throws AmbariException {
+    return clustersProvider.get().getCluster(prerequisiteCheck.getClusterName());
+  }
+
+  private Set<String> osFamiliesInCluster(Cluster cluster) {
```

Comment:
Javadoc

Reply:
Do we really need to add javadoc comments for private methods, especially short ones like these? I could have written one large method without using any privates and no one would have asked for javadocs in that case I suppose. IMHO if we require writing javadocs for every private methods it will discourage people to extract out logic into shorter method.

Reply:
We absolutely do. Every method should be properly documented. A developer should only ever need to read documentation to determine the inputs, what they mean, and what the method does/expects.

Reply:
I added the missing javadocs, but I still think it's not just unnecessary but also harmful to enforce javadoc at private method level because it encourages people to write large methods without factoring out anything.

Reply:
Just because you are writing a short description of what the method is doing and anything special that should be known about it (such as if it returns a copy of the underlying data set or an immutable one or if there are performance implications from calling it) should not preclude someone from writing concise methods. 

Documentation is always better than none. Ambari is riddled with classes with zero documentation where you need to examine each line to figure out what the heck is going on.

Reply:
> should not preclude someone from writing concise methods.

One of the reason why people like writing small methods is to eliminate the need of writing a bunch of comments. No one will do it if they need to write those comments anyway.

If someone implements some logic using one big method and an other person implements the same using 10 smaller methods then the amount of required comments will be completely different (1 versus 10), even if the logic is the same. So, this rule makes the amount of required comments highly style dependent.

Depending on the coding style, either people will complain about the missing comments even if the code is readable without them or they'll be completely fine with one unreadable "God method".

> Documentation is always better than none.

I disagree. It's a cost/benefit tradeoff. Comments need to be written and maintained and still they'll unavoidably go out of sync eventually, misleading the reader.

Many times they don't add any extra value (like most of the comments on getter/setters) or they're untrue or the method is so simple than reading the code doesn't take longer than reading the comment.

Reply:
> One of the reason why people like writing small methods is to eliminate the need of writing a bunch of comments. No one will do it if they need to write those comments anyway.

That is called being lazy. We shouldn't accept undocumented code because developers want to take shortcuts. There is no cost/benefit trade off here. Code needs to be clear, maintainable and understandable by anyone - even people looking at it for the first time. A method can be clean and concise, but should still have documentation. 

I don't know why we're arguing about this. No documentation is sloppy and lazy - period.

Reply:
> I don't know why we're arguing about this

We're not, because you're being dogmatic and not responding any of my arguments. So let's agree to disagree at least.

Reply:
I responded to your arguments directly by saying that asking developers to document methods should not lead to "god" methods. That even though smaller, concise methods are favorable, that should not preclude clear documentation so developers don't need to read every line of code to figure out what a method is or does.

Frankly, I am sticking to my guns here because Ambari is an undocumented mess in a lot of areas. I can open up a file at random and see _ZERO_ documentation in it. 

I'm sorry that you feel that spending the extra 20 seconds will lead to the unravelling of the universe, but in a project so large, we need to lead by example since it's too hard to enforce policies across the community. We can do this by having our code broken out into smaller, consumable methods that are still clearly documented.

Reply:
Ok, I still don't see why adding javadocs to 3 lines long private methods will solve any of the problems, but nevertheless I added those. I'm merging and closing this issue.